### ISSUE TITLE
Sorting isn't versioned correctly

### ISSUE THREAD

**Issue by ScopeyNZ** [silverstripe/silverstripe-versioned#180]
Sorting with versioned is not predictable and probably hasn't really been properly considered. Here's a few statements/examples to illustrate the issues:

### Problem

Currently when you sort a versioned object you may do one of three things:

#### 1) Always use `->write`
Update the object you are sorting with `->write()`, then update all objects in between the new and old positions (exclusive) by looping and also using `->write`.

__Observations:__
- This is slow - If you have 1000 elements and move position 1000 to position 1 you've got 1000 writes.
- You get a draft version of the object you updated and a draft version of all object in between - not great for usability
- Reverting to a version is ambiguous. When reverting, you'll have to revert some other objects too, but there's no way of determining which ones

#### 2) Use write only on the re-ordered object
Update the object you are sorting with `->write()`, update all other objects with an `UPDATE` query.

__Note:__ Elemental will do this (in master branch) and CMS already does this with SiteTree (although it still loops the objects and runs individual updates - I can raise an issue for that)

__Observations:__
- This is better for performance. You always have 2 queries
- Only the moved object has a draft version
- After moving and before publishing, you have some ambiguous sort. Consider 5 pages labelled 1 to 5. Move page 4 to after 1 (1 4 2 3 5). The sort on the published pages is (1 2 3 4 4) - pages 4 and 5 have the same sort. This means that the DB (especially PostgreSQL) might not return these in the expected order. This issue would be more prevalent with lots of unpublished sorting modifications.
- Reverting doesn't really work, because the order of the other versioned objects isn't corrected. You have a similar problem to the previous observation. This could be an "easy" fix though - you can treat the revert like a move and re-order all the elements between the old and new positions again.

#### 3) Direct update everything
Update all objects with direct queries
- It works but you get no versioning of sort order.

### Solution

I don't know.

There's some ideas I can think of (from simplest to hardest):
1. We go with option 3
2. We go with option 2, but we don't update the sort of anything until it's published - This would make it very hard to show draft sort order (but that's not currently considered anyway)
3. We store the sort order on some parent table (arbitrarily or by relation) meaning that sort orders can be viewed and reverted to. This would be the most powerful option.

**Comment by ScopeyNZ** [comment]
I'm also thinking if we define a way this should be done we can provide some API for it. 

In PHP: `$dataObject->move($afterObject)`
GraphQL: `mutation moveThing($ID, $AfterID)`
Apollo optimistic updates: (No example but we can have a generic-ish function that handles this so you get instant feedback)

**Comment by chillu** [comment]
Great write up! A few clarifications:

This is assuming `has_may`, right? A `many_many` can be a versioned relationship (many many through), presumably it doesn't have the same issues?

> update all other objects with an UPDATE query.

Could you detail what the query looks like? Sorry, too lazy to go hunting for it. Only found some logic in `BaseElement->Pos()` (https://github.com/dnadesign/silverstripe-elemental/blob/93f736fb9c68e2fe29d9f6514026b02cefaf1ade/src/Models/BaseElement.php#L904)

> After moving and before publishing, you have some ambiguous sort. Consider 5 pages labelled 1 to 5. Move page 4 to after 1 (1 4 2 3 5). The sort on the published pages is (1 2 3 4 4) - pages 4 and 5 have the same sort. 

I don't understand why changing the sort order on a draft record without publishing anything would change sort orders on any published pages?

> We go with option 2, but we don't update the sort of anything until it's published - This would make it very hard to show draft sort order 

Isn't this just a matter of `SELECT Sort FROM BaseElement_Live` vs. `SELECT Sort FROM BaseElement`?

**Comment by ScopeyNZ** [comment]
> This is assuming `has_may`, right? A `many_many` can be a versioned relationship (many many through), presumably it doesn't have the same issues?

You're right I didn't really consider `many_many`. If you versioned your sort through `many_many_through` you would still have the same problem. If you moved one object, you would have to update the sort on each "join" object, and now you have to choose when to publish each one of those. This would be even more confusing UX wise (I imagine) because the actual objects won't have any changes so I'm not sure how the user would know to publish their changed sort orders.

> Could you detail what the query looks like? Sorry, too lazy to go hunting for it. Only found some logic in `BaseElement->Pos()` ([dnadesign/silverstripe-elemental:src/Models/BaseElement.php@`93f736f`#L904](https://github.com/dnadesign/silverstripe-elemental/blob/93f736fb9c68e2fe29d9f6514026b02cefaf1ade/src/Models/BaseElement.php#L904))

Sure. You can find all objects between the old and new position of the moved object and either increment or decrement the sort order of them all. Consider items numbers 1 to 5 (1 2 3 4 5). Move 4 to the first position: (4 1 2 3 5). Before you move it, find all objects from 1st to 3rd positions (inclusive) and because we're moving block 4 to a higher position we must increment all of those (1 2 3 all get +1). 
Query: `UPDATE table SET Sort = Sort + 1 WHERE Sort >= 1 AND Sort <= 3`
Then: `$object->Sort = 1; $object->write()`.

Here's a snippet from a PR in elemental for seeing it in code: https://github.com/dnadesign/silverstripe-elemental/blob/034e100826fd4f349bd30cc4c62b8534d7f73b25/src/Services/ReorderElements.php#L63-L84

> I don't understand why changing the sort order on a draft record without publishing anything would change sort orders on any published pages?

That example was the idea where we only create a version for the moved object, meaning that other objects that will have their sort orders affected will be updated on the _live_ version. Referring back to the example:
> Move page 4 to after 1 (1 4 2 3 5). The sort on the published pages is (1 2 3 4 4) - pages 4 and 5 have the same sort.
It's actually incorrect, the sort on the published pages will be (1 3 4 4 5). It might make more sense if I list the _live_ sort order on each of the objects at the end:

| Object | Sort |
| --- | --- |
| 1 | 1 |
| 2 | 3 |
| 3 | 4 |
| 4 | 4 |
| 5 | 5 |

All the objects other than 4 (the one that was actually moved) have been updated to make way for the sort order of 4 that will be published, but it's not published yet!

> Isn't this just a matter of `SELECT Sort FROM BaseElement_Live` vs. `SELECT Sort FROM BaseElement`?

If option 2 was chosen, but we don't update sort orders of live blocks until publish time then we'd have to calculate what the updated order of non-moved blocks would be. Another example:

Initial list: A B C D E F G H I J
Move I to beginning: I A B C D E F G H J
Move B to end: I A C D E F G H J B
Move H to after A: I A H C D E F G J B
Move I again: A H C D E I F G J B
Final sort values (only moved objects have been updated): 

| Letter | Sort | Comment |
| --- | --- | --- |
| A | 1 | |
| B | 10 | |
| C | 3 | |
| D | 4 | |
| E | 5 | |
| F | 6 | F was never moved - still has original sort |
| G | 7 | |
| H | 3 | H was moved to 3, even though it should be at 2 now |
| I | 6 | Clashes with F now |
| J | 10 | J was never moved, now has same order as B |

Working through this example I'm realising that we can't update sort of everything else on publish, because there's a LOT of ambiguity when you have many sorts one after the other.

Sorry!

**Comment by chrispenny** [comment]
# Versioning with vanilla SS4.3, Elemental, and Elemental List

## Installation

```
"silverstripe/recipe-cms": "4.3.x-dev",
"dnadesign/silverstripe-elemental": "3.x-dev",
"dnadesign/silverstripe-elemental-list": "1.x-dev as 1.1.0",
```

## Process (text)

**Set up Page content**
3 x ElementList, each with 1 x ElementContent

**Click History Tab**
There is only one version (with no Blocks)

**Update any Metadata field**
**Save page**
**Click History Tab**
There is still only one version (with no Blocks)

**Refresh the Page**
You can now see additional versions

**Update Block text for each `ElementContent`**
There is no indicator that a save or publish is required

**Change Block order**
Indicator to publish now triggers
There is no "Cancel draft changes" option available

**Update a Metadata field**
Now the indicator is visible
Now there is a "Cancel draft changes" option

**Cancel draft changes**
Now worked as expected

## Process (images)

![1](https://user-images.githubusercontent.com/505788/47972286-ab4c8080-e0ff-11e8-9a69-d17f48edcba3.png)
![2](https://user-images.githubusercontent.com/505788/47972287-abe51700-e0ff-11e8-84e8-2c1910be7b55.png)
![3](https://user-images.githubusercontent.com/505788/47972288-abe51700-e0ff-11e8-89b0-2bc378eaedf6.png)
![4](https://user-images.githubusercontent.com/505788/47972289-abe51700-e0ff-11e8-8f5f-8b8e292181ed.png)
![5](https://user-images.githubusercontent.com/505788/47972290-abe51700-e0ff-11e8-8017-7042235419c2.png)
![6](https://user-images.githubusercontent.com/505788/47972291-ac7dad80-e0ff-11e8-8aea-096b30af1ae7.png)
![7](https://user-images.githubusercontent.com/505788/47972292-ac7dad80-e0ff-11e8-8edf-359178f0f885.png)
![8](https://user-images.githubusercontent.com/505788/47972293-ad164400-e0ff-11e8-8729-7f902f6e341a.png)
![9](https://user-images.githubusercontent.com/505788/47972294-adaeda80-e0ff-11e8-89b1-c4d2f5f4da50.png)
![10](https://user-images.githubusercontent.com/505788/47972295-adaeda80-e0ff-11e8-812d-772d084934b2.png)
![11](https://user-images.githubusercontent.com/505788/47972296-adaeda80-e0ff-11e8-927b-33e168b8e2a5.png)
![12](https://user-images.githubusercontent.com/505788/47972297-ae477100-e0ff-11e8-94ef-fc9b1aa30536.png)
![13](https://user-images.githubusercontent.com/505788/47972298-ae477100-e0ff-11e8-82f1-10352f1ac66e.png)
![14](https://user-images.githubusercontent.com/505788/47972299-ae477100-e0ff-11e8-8259-3785bab2bbc9.png)
![15](https://user-images.githubusercontent.com/505788/47972300-aee00780-e0ff-11e8-82a4-027b370dc0f7.png)
![16](https://user-images.githubusercontent.com/505788/47972301-aee00780-e0ff-11e8-8953-a7e6e9c43474.png)
![17](https://user-images.githubusercontent.com/505788/47972302-af789e00-e0ff-11e8-99b7-49642229825c.png)
![18](https://user-images.githubusercontent.com/505788/47972303-af789e00-e0ff-11e8-8c39-009cb9dd2211.png)

**Comment by chrispenny** [comment]
# Versioning with vanilla SS4.3, Fluent, Elemental, and Elemental List

## Installation

```
"silverstripe/recipe-cms": "4.3.x-dev",
"dnadesign/silverstripe-elemental": "3.x-dev",
"dnadesign/silverstripe-elemental-list": "1.x-dev as 1.1.0",
"tractorcow/silverstripe-fluent": "4.1.x-dev"
```

## Process (text)

** Set up Locales**
International (en_NZ), default_locale
Japan (ja_JP)

No fallback Locales for either.

**Set up Page content**
International has **no** content

Japan has the following content:
3 x ElementList, each with 1 x ElementContent (this should be "version 1")

**From the Japan Locale**

**Re-order the Blocks**
**Update Block text for each `ElementContent` (this should be "version 2")**
**Update any page metadata**
**Save & Publish**

**Re-order the Blocks again**
**Update Block text for each `ElementContent` (this should be "version 3")**
**Update any page metadata**
**Save & Publish**

**Click History Tab**
**Click on the "Version 2" record**
Record looks correct

**Revert to this version**
All of the Blocks are still on "version 3" (content, and sort order), but the page metadata has been reverted

## Process (images)

See the end for a copy of `Page.php` containing these customisations.
![0](https://user-images.githubusercontent.com/505788/47972399-d71c3600-e100-11e8-88be-793d2eddab20.png)
![1](https://user-images.githubusercontent.com/505788/47972400-d71c3600-e100-11e8-8b04-a7fd5ccfa5da.png)
![2](https://user-images.githubusercontent.com/505788/47972401-d71c3600-e100-11e8-87e0-8e12431ed280.png)
![3](https://user-images.githubusercontent.com/505788/47972402-d7b4cc80-e100-11e8-9f9f-18e170e29880.png)
![4](https://user-images.githubusercontent.com/505788/47972403-d7b4cc80-e100-11e8-8776-93221e3ded86.png)
![5](https://user-images.githubusercontent.com/505788/47972404-d7b4cc80-e100-11e8-8eab-771c57459f9d.png)
![6](https://user-images.githubusercontent.com/505788/47972405-d84d6300-e100-11e8-9416-573e70fae906.png)
![7](https://user-images.githubusercontent.com/505788/47972406-d84d6300-e100-11e8-83d5-6fb45cbc5297.png)
![8](https://user-images.githubusercontent.com/505788/47972407-d84d6300-e100-11e8-95f5-0e1135598ac0.png)
![9](https://user-images.githubusercontent.com/505788/47972408-d8e5f980-e100-11e8-84f9-7401f0d167f1.png)
![10](https://user-images.githubusercontent.com/505788/47972409-d97e9000-e100-11e8-9247-19f3c80e0afa.png)
![11](https://user-images.githubusercontent.com/505788/47972410-d97e9000-e100-11e8-974e-44535bd00c4b.png)
![12](https://user-images.githubusercontent.com/505788/47972411-d97e9000-e100-11e8-9d32-1cac88778479.png)
![13](https://user-images.githubusercontent.com/505788/47972412-d97e9000-e100-11e8-86e2-1e418914dd4a.png)
![14](https://user-images.githubusercontent.com/505788/47972413-da172680-e100-11e8-8ebd-b1b81aae160b.png)
![15](https://user-images.githubusercontent.com/505788/47972414-da172680-e100-11e8-8958-fe9adeda6483.png)
![16](https://user-images.githubusercontent.com/505788/47972415-daafbd00-e100-11e8-9137-d55defd5b768.png)
![17](https://user-images.githubusercontent.com/505788/47972416-daafbd00-e100-11e8-97d1-519f93f43b1c.png)
![18](https://user-images.githubusercontent.com/505788/47972417-daafbd00-e100-11e8-8601-24b722d1e6b3.png)
![19](https://user-images.githubusercontent.com/505788/47972418-db485380-e100-11e8-9621-f815eb58cc3b.png)
![20](https://user-images.githubusercontent.com/505788/47972419-dbe0ea00-e100-11e8-9579-824f1cc9287d.png)
![21](https://user-images.githubusercontent.com/505788/47972420-dbe0ea00-e100-11e8-8a2f-5d23ad8fb64f.png)

```php
<?php

namespace {

    use DNADesign\Elemental\Extensions\ElementalPageExtension;
    use DNADesign\Elemental\Models\ElementalArea;
    use SilverStripe\CMS\Model\SiteTree;
    use TractorCow\Fluent\Extension\FluentVersionedExtension;

    /**
     * Class Page
     *
     * @property int ElementalAreaID
     * @method ElementalArea ElementalArea()
     * @mixin FluentVersionedExtension
     * @mixin ElementalPageExtension
     */
    class Page extends SiteTree
    {
        /**
         * @var array
         */
        private static $extensions = [
            ElementalPageExtension::class,
        ];

        /**
         * @var array
         */
        private static $db = [
            'TextField' => 'HTMLText',
        ];

        /**
         * @var array
         */
        private static $has_one = [];

        /**
         * Fluent specific configuration
         *
         * @var array
         */
        private static $field_include = [
            'ElementalAreaID',
        ];

        public function onBeforeWrite() // phpcs:ignore SlevomatCodingStandard.TypeHints
        {
            parent::onBeforeWrite();

            // if localised page instance is being created proceed to duplicate elemental data
            // NOTE clone elemental only if this page already exists otherwise there is nothing to clone
            if (!$this->isDraftedInLocale() && $this->isInDB()) {
                $elementalArea = $this->ElementalArea();

                $sourceElementalArea = ElementalArea::get()->byID($this->forceElementalAreaId);
                if (!empty($sourceElementalArea)) {
                    $elementalArea = $sourceElementalArea;
                }

                $elementalAreaNew = $elementalArea->duplicate();
                $this->ElementalAreaID = $elementalAreaNew->ID;
            }

            return;
        }
    }
}
```

**Comment by ScopeyNZ** [comment]
I've unassigned @micmania1 because the work he did was captured in other issues & PRs. This issue really refers to the high-level how is sorting supposed to be managed with versioned objects.

I think the issue got a bit carried away - perhaps the title should be more descriptive.

**Comment by ScopeyNZ** [comment]
Actually because it got a bit side tracked I'm going to re-raise the issue if that's okay with you @chillu? 

I'd really like the goal of this ticket to be a solution to this problem statement:

"I want to sort a versioned object that belongs to a parent object, and for it to make sense and work out of the box".

Currently all our instances of this already (`SiteTree`, `BaseElement` and `File`) all do it their own way - and they're all flawed.

**Comment by ScopeyNZ** [comment]
Okay in the end I've just hidden @chrispenny's comments. That was some stellar investigation and we've identified this as a pretty big issue in https://github.com/silverstripe/silverstripe-versioned/issues/195.

I've update the OP with a few examples and some possible solutions. I'm not sure about the actual impact of this because it's been like this since the dawn of versioning.

I've tagged this as needing feedback from @silverstripe/core-team because hopefully we can at least get consensus on a potential solution...

**Comment by sminnee** [comment]
I'd say that in the first instance we want to close off #195 as best we can without this, acknowledge any limitations, and merge it into `4` so we can at least bank _some_ improvement before getting lost in this one.

Secondly, it occurs to me that what we might need to do is create "the ordering of a set of siblings" as a separate piece of data for the purposes of versioning.

So you have, say, `SiteTree_SortVersion` as a separate table. When you change the order of some items, the Sort values of all the sibling items get logged in this table rather than as any version changes of those siblings. It ends up being managed in a similar way to how we're managing other related data objects.

Then, when you need to get historic data of SiteTree, you join SiteTree_SortVersion and grab the Sort value form there, rather than using SiteTree_Versions.Sort.

It's just a high-level sketch but this might be more robust?

To allow for sufficient backward compatibility, I would probably retain the existence of the SiteTree_Versions.Sort field as-is, and maybe have a build task or dev/build book to build out the initial version of the SiteTree_SortVersion dataset.

This would probably require some special knowledge added to the system to say "oh this field is for sorting". The improved behaviour would only apply if this feature was activated. We could enable it for SiteTree and blocks, but others would need to add it manually. As long as the system worked with its current bugs if you didn't do that, that seems fine to introduce in a minor release.

**Comment by ScopeyNZ** [comment]
> I'd say that in the first instance we want to close off #195 as best we can without this

Yes to be clear I consider this an entirely different issue. #195 is about tracking changes across an relationship graph. This is about sorting individual objects and only involves siblings when talking about how they need to be updated when moving an object around. Mention of "parents" is just as a potential solution. I don't think we should relate this to capturing changes in the "relationship graph" - #195.

> So you have, say, SiteTree_SortVersion as a separate table

Having Sort on a different table isn't a bad idea from my PoV but I think it still makes more sense to have the sort against a parent - If you think about moving a block around you don't really expect to look at the history of that block to see it's movement within a list, you'd look at the page (or "elemental area") to see how things were re-ordered.

The problem with this is that there's nothing that's versioned that's a "parent" to `SiteTree` - `SiteTree`'s only have a "parent" when subsites is installed. You almost need a new dataobject `SiteTreeSort` with `SiteTreeSort_Live`.

> I would probably retain the existence of the SiteTree_Versions.Sort field as-is

This kind of scares me. I think you're right so far as preventing a fatal on trying to update non-existent fields - I'm sure there's plenty of code in the wild that's `UPDATE SiteTree SET sort = ?`. It'd be nice to hook into `onBeforeWrite` and catch updates to Sort in that way though.

I still am thinking of this like an extension that needs to be configured in some way:

```
SilverStripe\CMS\Model\SiteTree
  extensions:
    ordering: SilverStripe\Versioned\Sortable
  sort:
    [some config here...]
```

And then you can:

```
$page = SiteTree::get()->byId(1);
$page->moveAfter(SiteTree::get()->byId(9)) // or just ->moveAfter(9)
```

**Comment by chillu** [comment]
So there's four kinds of sort (assuming versioned objects on both sides of the relationship):

 * Sort Variation 1: `many_many_extraFields`: Legacy, should ignore
 * Sort Variation 2: `has_many` sort (e.g. `ElementArea has_many Element`, `SiteTree` with parents)
 * Sort Variation 3: Sort without relationship (e.g. toplevel `SiteTree`)
 * Sort Variation 4: `many_many_through`: Recommended for versioning `many_many`, has same issue. Conceptually the same as Sort Variation 2

Only Sort Variation 3 actually requires sorts to be stored on the record itself, because there's no relationship owner on the other side. We've somewhat abused `SiteConfig` for "toplevel SiteTree" config before (view/edit permissions), and it's already a hard dependency on `silverstripe/cms`.

I'm a bit worried about table explosion if we introduce `<sorted-record>_SortVersion` for everything. `SiteTree_Sort` and `Element_Sort` are just the tip of the iceberg. We'd also get `GalleryItem_Sort`, `Link_Sort`, `RelatedDocument_Sort`, potentially dozens of tables on a large install. Not only because the schema gets unwieldy, but because that's all stuff we need to join on to generate already expensive rollbacks and history views (#195 won't solve that).

How about making `Sort` a property on the owner of the sorted relationship. So a new `ElementalArea.ElementSorts` column, which contains a map of element identifiers to their sort order. That column is written on draft and live stages by whatever modifies the relationship (usually GridField), and updates the full list of siblings. We'd take away the ability to just call `$myRecord->update(['Sort' => 2])->write()`. This would also address the fact that records can have more than one sort property since they can be part of more than one `has_many` relationship. This field would need some serialisation, and support for in-memory sorting on the ORM level. That's really not ideal for paging through long sibling lists (e.g. entries under a "blog holder"), or limiting nodes like in our tree view. So ... a non-starter.

Aside: We could make this column a [JSON type](https://dev.mysql.com/doc/refman/5.7/en/json.html), but that would require MySQL 5.7 (see [discussion](https://github.com/silverstripe/silverstripe-framework/issues/8651)). And you'd need MySQL 8.0 for proper [JSON table functions](https://dev.mysql.com/doc/refman/8.0/en/json-table-functions.html) to do the required virtual table joins, so it's a non starter as well. We could just support sorting on Postgres? ;)

I'm really tempted to just say "sort isn't versioned". We've spent an enormous amount of time on versioning edge cases in the last years. Would that really be so bad for author experience? /cc @clarkepaul

**Comment by sminnee** [comment]
> How about making Sort a property on the owner of the sorted relationship. So a new ElementalArea.ElementSorts column, which contains a map of element identifiers to their sort order. 

So you're recommending an array field over a separate table? How confident are we that this will make faster rather than slower queries?

**Comment by chillu** [comment]
I like framing this as optional partitioning, makes a lot of sense. 

"separate table vs. array field": My first concern is amount of tables. I've seen projects with 450 tables, and think SilverStripe is already pretty insane in terms of getting your head around the data model of your site as an average developer. It also increases dev/build times (more to inspect and adjust). Speed of array fields depends on database implementation - I'd wager it's just as fast as joins in Postgres, and could be slower in MySQL. it's a non-starter anyway, since we'd need to bump MySQL support to 8.0 for that. It's definitely slower than joins if we do this in-memory. Can we get a way with one or two tables for sort versioning? `SortVersion` and `SortVersion_Live`, with cols for `ObjectClass` etc? Am I the only one that's freaked out by projects with hundreds of database tables? :D

To clarify, an "unversioned sort" implementation really means "synced sort between stages". So when you change sort of an item, it directly writes every draft and every live record in the same partition without creating a new version. This introduces new issues with rollback though: You'd have to reset all items in the same partition to the date point in time (closest version). That might be just as time intensive to implement as actually fixing sort versioning in the first place. Aaaaargh.

**Comment by sminnee** [comment]
> Can we get a way with one or two tables for sort versioning? SortVersion and SortVersion_Live, with cols for ObjectClass etc?

If ObjectClass was an enum included in the index that would probably be fine.

> Am I the only one that's freaked out by projects with hundreds of database tables?

It's probably worth clarifying that actual downsides of it, beyond heebie-jeebies. I've tended to recognise that heebie-jeebies can be a sign that something is problematic, but equally that something is merely unusual.

**Comment by tractorcow** [comment]
I had a similar problem with fluent localisation; The issue was I needed the ability to whitelist which fields needed localisation (or in this case, versioning) and not localise others.

Similarly, I had separate tables, except that the `_Localised` table contained only columns requiring localisation. Those missing fields were joined from the base table on every query.

Perhaps what versioning needs is a way to whitelist which fields are versioned, conditionally joining the base table when loading / lazy-loading fields that are not?

In other words, `Sort` is not versioned, and wouldn't even be a column on `_Live` rows.

We've essentially dropped the concept of `on live but not on stage` so I'm not worried about missing joins (stage is source of truth).

I prefer joining existing tables (and trimming the complexity of one of them) much more than adding yet another table to the schema.

**Comment by chillu** [comment]
Flagging that we've spent literally hundreds of hours on versioning behaviour in the Open Sourcerers team in the last three months, so from my team's perspective I'm going to park this. It's a long standing issue, which can be remedied from an author perspective by simply publishing everything to get it back into a consistent state.

**Comment by sminnee** [comment]
Would it be worth noting the quirk in our user and/or dev docs?
